File: apps/web/src/lib/api/competitors.ts
/**
 * Competitors API — competitor intelligence dashboard.
 */

import { apiFetch } from "./client";

// ── Types ────────────────────────────────────────────────

export interface Competitor {
  id: string;
  user_id: string;
  brand_id?: string | null;
  name: string;
  platform: string;
  profile_url: string;
  positioning?: string | null;
  niche?: string | null;
  estimated_followers?: number | null;
  pricing_tier?: string | null;
  notes?: string | null;
  threat_level: number;
  threat_level_override?: boolean;
  status: string;
  created_at?: string | null;
  updated_at?: string | null;
  latest_metrics?: CompetitorMetric | null;
  metrics_history?: CompetitorMetric[];
  recent_content?: CompetitorContent[];
}

export interface CompetitorMetric {
  id?: string;
  competitor_id?: string;
  recorded_at?: string | null;
  followers?: number | null;
  engagement_rate?: number | null;
  post_frequency_weekly?: number | null;
  avg_post_engagement?: number | null;
  top_topic?: string | null;
  source?: string;
  created_at?: string | null;
}

export interface CompetitorContent {
  id?: string;
  competitor_id?: string;
  published_at?: string | null;
  platform?: string | null;
  title?: string | null;
  url?: string | null;
  content_preview?: string | null;
  topics?: string[];
  engagement_count?: number | null;
  engagement_rate?: number | null;
  format?: string;
  created_at?: string | null;
}

export interface CompetitorComparison {
  competitor_id: string;
  competitor_name: string;
  user_metrics: Record<string, any>;
  competitor_metrics: Record<string, any>;
  insights: string[];
}

export interface CompetitorAnalysisReport {
  competitor_id: string;
  summary: string;
  strengths: string[];
  weaknesses: string[];
  content_pillars: string[];
  threat_assessment: string;
}

export interface ContentGap {
  topic: string;
  covered_by_competitors: string[];
  your_coverage: boolean;
  priority: string;
}

export interface ContentGapAnalysis {
  gaps: ContentGap[];
  your_unique_topics: string[];
  shared_topics: string[];
}

export interface CompetitorCreateData {
  name: string;
  platform?: string;
  profile_url: string;
  positioning?: string;
  niche?: string;
  pricing_tier?: string;
  notes?: string;
  threat_level?: number;
  brand_id?: string;
}

export interface CompetitorUpdateData {
  name?: string;
  platform?: string;
  profile_url?: string;
  positioning?: string;
  niche?: string;
  pricing_tier?: string;
  notes?: string;
  threat_level?: number;
  status?: string;
}

export interface ThreatScoreDetail {
  calculated_score: number;
  engagement_growth_factor: number;
  content_overlap_factor: number;
  frequency_factor: number;
  follower_ratio_factor: number;
  reasoning: string;
  is_overridden: boolean;
}

export interface CompetitorAlert {
  id?: string;
  competitor_id: string;
  competitor_name: string;
  alert_type: string;
  detail: string;
  metric_before?: number | null;
  metric_after?: number | null;
  severity: string;
  created_at?: string | null;
}

export interface IntelligenceFeedItem {
  item_type: string;
  competitor_id: string;
  competitor_name: string;
  summary: string;
  threat_level?: number | null;
  date?: string | null;
}

export interface IntelligenceFeed {
  active_competitors: number;
  avg_threat_level: number;
  latest_analysis_date?: string | null;
  open_alerts: number;
  recent_analyses: IntelligenceFeedItem[];
  recent_alerts: CompetitorAlert[];
  benchmarks: Record<string, any>;
}

// ── Platform options ────────────────────────────────────

export const PLATFORMS = [
  { value: "linkedin", label: "LinkedIn" },
  { value: "twitter", label: "Twitter/X" },
  { value: "youtube", label: "YouTube" },
  { value: "tiktok", label: "TikTok" },
  { value: "instagram", label: "Instagram" },
  { value: "website", label: "Website" },
  { value: "other", label: "Other" },
] as const;

export const THREAT_LEVELS = [
  { value: 1, label: "Low", color: "text-green-400" },
  { value: 2, label: "Minor", color: "text-green-300" },
  { value: 3, label: "Moderate", color: "text-yellow-400" },
  { value: 4, label: "High", color: "text-orange-400" },
  { value: 5, label: "Critical", color: "text-red-400" },
] as const;

// ── API Methods ─────────────────────────────────────────

export const competitorsApi = {
  list: (params?: { brand_id?: string; status?: string }) => {
    const qs = new URLSearchParams();
    if (params?.brand_id) qs.set("brand_id", params.brand_id);
    if (params?.status) qs.set("status", params.status);
    const query = qs.toString();
    return apiFetch<Competitor[]>(`/competitors${query ? `?${query}` : ""}`);
  },

  create: (data: CompetitorCreateData) =>
    apiFetch<Competitor>("/competitors", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  get: (id: string) => apiFetch<Competitor>(`/competitors/${id}`),

  update: (id: string, data: CompetitorUpdateData) =>
    apiFetch<Competitor>(`/competitors/${id}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    }),

  remove: (id: string) =>
    apiFetch<void>(`/competitors/${id}`, { method: "DELETE" }),

  getMetrics: (id: string, days = 30) =>
    apiFetch<CompetitorMetric[]>(`/competitors/${id}/metrics?days=${days}`),

  recordMetrics: (id: string, data: Partial<CompetitorMetric>) =>
    apiFetch<CompetitorMetric>(`/competitors/${id}/metrics`, {
      method: "POST",
      body: JSON.stringify(data),
    }),

  getContent: (id: string, limit = 20) =>
    apiFetch<CompetitorContent[]>(
      `/competitors/${id}/content?limit=${limit}`,
    ),

  refresh: (id: string) =>
    apiFetch<{ competitor_id: string; content_items_added: number }>(
      `/competitors/${id}/refresh`,
      { method: "POST" },
    ),

  analyze: (id: string) =>
    apiFetch<CompetitorAnalysisReport>(`/competitors/${id}/analyze`, {
      method: "POST",
    }),

  compare: (competitorId: string) =>
    apiFetch<CompetitorComparison>(
      `/competitors/comparison?competitor_id=${competitorId}`,
    ),

  getGaps: (brandId?: string) => {
    const qs = brandId ? `?brand_id=${brandId}` : "";
    return apiFetch<ContentGapAnalysis>(`/competitors/gaps${qs}`);
  },

  getIntelligenceFeed: (brandId?: string) => {
    const qs = brandId ? `?brand_id=${brandId}` : "";
    return apiFetch<IntelligenceFeed>(`/competitors/intelligence${qs}`);
  },

  getAlerts: (limit = 20) =>
    apiFetch<CompetitorAlert[]>(`/competitors/alerts?limit=${limit}`),

  triggerFullAnalysis: (brandId?: string) =>
    apiFetch<{ analyzed: number; results: CompetitorAnalysisReport[] }>(
      "/competitors/full-analysis",
      {
        method: "POST",
        body: JSON.stringify(brandId ? { brand_id: brandId } : {}),
      },
    ),
};

File: apps/api/app/schemas/competitors.py
"""Pydantic schemas for the Competitor Intelligence system."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field, field_validator


VALID_PLATFORMS = {
    "linkedin", "twitter", "youtube", "tiktok", "instagram", "website", "other",
}

VALID_CONTENT_FORMATS = {
    "post", "video", "carousel", "thread", "story", "article", "other",
}


# ── Create / Update ────────────────────────────────────────────

class CompetitorCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    platform: str = Field("website")
    profile_url: str = Field(..., min_length=1, max_length=2000)
    positioning: Optional[str] = Field(None, max_length=1000)
    niche: Optional[str] = Field(None, max_length=200)
    pricing_tier: Optional[str] = Field(None, max_length=100)
    notes: Optional[str] = Field(None, max_length=5000)
    threat_level: int = Field(3, ge=1, le=5)
    brand_id: Optional[str] = None

    @field_validator("platform")
    @classmethod
    def validate_platform(cls, v: str) -> str:
        if v not in VALID_PLATFORMS:
            raise ValueError(f"Invalid platform: {v}. Valid: {sorted(VALID_PLATFORMS)}")
        return v


class CompetitorUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    platform: Optional[str] = None
    profile_url: Optional[str] = Field(None, min_length=1, max_length=2000)
    positioning: Optional[str] = Field(None, max_length=1000)
    niche: Optional[str] = Field(None, max_length=200)
    pricing_tier: Optional[str] = Field(None, max_length=100)
    notes: Optional[str] = Field(None, max_length=5000)
    threat_level: Optional[int] = Field(None, ge=1, le=5)
    brand_id: Optional[str] = None
    status: Optional[str] = None

    @field_validator("platform")
    @classmethod
    def validate_platform(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in VALID_PLATFORMS:
            raise ValueError(f"Invalid platform: {v}. Valid: {sorted(VALID_PLATFORMS)}")
        return v

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in ("active", "archived"):
            raise ValueError("status must be 'active' or 'archived'")
        return v


# ── Output ──────────────────────────────────────────────────────

class CompetitorMetricSnapshot(BaseModel):
    followers: Optional[int] = None
    engagement_rate: Optional[float] = None
    post_frequency_weekly: Optional[float] = None
    avg_post_engagement: Optional[int] = None
    top_topic: Optional[str] = None


class CompetitorOut(BaseModel):
    id: str
    user_id: str
    brand_id: Optional[str] = None
    name: str
    platform: str
    profile_url: str
    positioning: Optional[str] = None
    niche: Optional[str] = None
    estimated_followers: Optional[int] = None
    pricing_tier: Optional[str] = None
    notes: Optional[str] = None
    threat_level: int = 3
    threat_level_override: bool = False
    status: str = "active"
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    latest_metrics: Optional[CompetitorMetricSnapshot] = None


# ── Metrics ─────────────────────────────────────────────────────

class CompetitorMetricRecord(BaseModel):
    followers: Optional[int] = None
    engagement_rate: Optional[float] = Field(None, ge=0, le=100)
    post_frequency_weekly: Optional[float] = Field(None, ge=0)
    avg_post_engagement: Optional[int] = Field(None, ge=0)
    top_topic: Optional[str] = Field(None, max_length=200)
    source: str = Field("manual")


class CompetitorMetricOut(CompetitorMetricRecord):
    id: str
    competitor_id: str
    recorded_at: Optional[datetime] = None
    created_at: Optional[datetime] = None


# ── Content ─────────────────────────────────────────────────────

class CompetitorContentRecord(BaseModel):
    published_at: Optional[datetime] = None
    platform: Optional[str] = None
    title: Optional[str] = Field(None, max_length=500)
    url: Optional[str] = Field(None, max_length=2000)
    content_preview: Optional[str] = Field(None, max_length=5000)
    topics: List[str] = Field(default_factory=list)
    engagement_count: Optional[int] = Field(None, ge=0)
    engagement_rate: Optional[float] = Field(None, ge=0)
    format: str = Field("post")

    @field_validator("format")
    @classmethod
    def validate_format(cls, v: str) -> str:
        if v not in VALID_CONTENT_FORMATS:
            raise ValueError(f"Invalid format: {v}. Valid: {sorted(VALID_CONTENT_FORMATS)}")
        return v


class CompetitorContentOut(CompetitorContentRecord):
    id: str
    competitor_id: str
    created_at: Optional[datetime] = None


# ── Comparison & Analysis ───────────────────────────────────────

class CompetitorComparison(BaseModel):
    competitor_id: str
    competitor_name: str
    user_metrics: Dict[str, Any] = Field(default_factory=dict)
    competitor_metrics: Dict[str, Any] = Field(default_factory=dict)
    insights: List[str] = Field(default_factory=list)


class CompetitorAnalysisReport(BaseModel):
    competitor_id: str
    summary: str
    strengths: List[str] = Field(default_factory=list)
    weaknesses: List[str] = Field(default_factory=list)
    content_pillars: List[str] = Field(default_factory=list)
    threat_assessment: str = ""


class ContentGap(BaseModel):
    topic: str
    covered_by_competitors: List[str] = Field(default_factory=list)
    your_coverage: bool = False
    priority: str = "medium"


class ContentGapAnalysis(BaseModel):
    gaps: List[ContentGap] = Field(default_factory=list)
    your_unique_topics: List[str] = Field(default_factory=list)
    shared_topics: List[str] = Field(default_factory=list)


# ── Threat Scoring ────────────────────────────────────────────

VALID_ALERT_TYPES = {
    "follower_surge", "engagement_drop", "positioning_shift",
    "content_spike", "new_strategy",
}

VALID_ALERT_SEVERITIES = {"low", "medium", "high"}

# Threat scoring weights (must sum to 1.0)
THREAT_WEIGHTS = {
    "engagement_growth": 0.30,
    "content_overlap": 0.25,
    "frequency": 0.25,
    "follower_ratio": 0.20,
}


class ThreatScoreDetail(BaseModel):
    """Dynamic threat score calculation breakdown."""
    calculated_score: float = Field(..., ge=1.0, le=5.0)
    engagement_growth_factor: float = 0.0
    content_overlap_factor: float = 0.0
    frequency_factor: float = 0.0
    follower_ratio_factor: float = 0.0
    reasoning: str = ""
    is_overridden: bool = False


class CompetitorAlert(BaseModel):
    """A competitor intelligence alert."""
    id: Optional[str] = None
    competitor_id: str
    competitor_name: str
    alert_type: str
    detail: str
    metric_before: Optional[float] = None
    metric_after: Optional[float] = None
    severity: str = "medium"
    created_at: Optional[str] = None

    @field_validator("alert_type")
    @classmethod
    def validate_alert_type(cls, v: str) -> str:
        if v not in VALID_ALERT_TYPES:
            raise ValueError(f"Invalid alert_type: {v}. Valid: {sorted(VALID_ALERT_TYPES)}")
        return v

    @field_validator("severity")
    @classmethod
    def validate_severity(cls, v: str) -> str:
        if v not in VALID_ALERT_SEVERITIES:
            raise ValueError(f"Invalid severity: {v}. Valid: {sorted(VALID_ALERT_SEVERITIES)}")
        return v


# ── Intelligence Feed ─────────────────────────────────────────

class IntelligenceFeedItem(BaseModel):
    """Single item in the intelligence feed."""
    item_type: str  # "analysis" | "alert"
    competitor_id: str
    competitor_name: str
    summary: str
    threat_level: Optional[int] = None
    date: Optional[str] = None


class IntelligenceFeed(BaseModel):
    """Aggregated competitor intelligence for the feed page."""
    active_competitors: int = 0
    avg_threat_level: float = 0.0
    latest_analysis_date: Optional[str] = None
    open_alerts: int = 0
    recent_analyses: List[IntelligenceFeedItem] = Field(default_factory=list)
    recent_alerts: List[CompetitorAlert] = Field(default_factory=list)
    benchmarks: Dict[str, Any] = Field(default_factory=dict)
